README.md

```diff
@@ -10,260 +10,6 @@ This crate provides basic imaging processing functions and methods for convertin
 
 All image processing functions provided operate on types that implement the `GenericImage` trait and return an `ImageBuffer`.
 
-## 1. Documentation
+## Documentation
```

Comment:
The freestanding comment was probably too vague. While there is value in the root library documentation, moving all of it into the library does not seem right either and the motivation for doing so was at least partially factually incorrect in that it does not contribute to its correctness. This would go smoother if there was some pros and cons where to place which parts. I'm open to having a discussion on this, it surely was not prioritized for a long time. As a programmer, the Readme should present some argument for including the library and having _no_ code in it does not appear too convincing to me, so maybe we can preserve some of it, it _is_ tested afterall?

Let's get to the rest of the document. I'm confident we can increase the value even more by splitting it into chunks of different purpose. For example, the table of supported formats is likely misplaced _within_ the library as there we already have the list of modules and implementors of `ImageDecoder` which is more accurate for the particular selected set of features and would be a better place for additional information overall. But it clearly is useful in the Readme as it quickly shows if the library is fit for a particular purpose before it is added to a project.

The traits and structures are probably better served by simply linking their documentation, it's technical information and doesn't add much for judging general applicability. However, I don't think it adds much to the crate documentation either, a condensed paragraph linking to the correct modules and traits would be clearer (and moving relevant portions into the documentation of the traits themselves). The `Pixel` section is plainly outdated and  as well better represented if the `impl` section of the trait were simply linked, simiarly for `GenericImage`. The `ImageBuffer` and `DynamicImage` sections could similarly be moved to the structs themselves and linked in a short, high-level paragraph.

I would not want to remove the fractal image from the Readme, somewhat for sentimental value. The code example may be too much for a Readme but it does not seem very applicable to the generic library documentation either. Since `examples` are not included in the current packed crate we can't simply link to the example either, or at least need to include them. That would be possible.

Reply:
Yep I'm happy to work on this to try to get it to a great place. That said, I'm not sure how much time I'll have to work on this so let's try to come up with a _series_ of changes which I can put into separate incremental PRs as I have time.

A few opinions on the docs as-is:

* They're really good!
* Specifically, the code examples are the right level of complexity and I think are really helpful for newbies to the crate. (I'd definitely want to keep at least one simple and one complex example in the top-level documentation, e.g. the open and save example and the fractal example)
* I think the explanation of the roles of `GenericImage`/`ImageBuffer`/`DynamicImage` is good and belongs in the top level documentation too. Yes, it could be shorter, but I don't see this as urgent.
* I think the table of formats is actually beneficial and probably belongs in the top-level documentation. I think it shows the number one feature of this library which is to support multiple image formats. (I'm not sure if you'd agree that's the number one feature but I assume it's why a lot of people want to use `image`).
* I too really like the fractal :)
* I think it would be shame to _split_ information across the `README.md` and `src/lib.rs`. I think it's useful having a single page with all the information.

So in short - I didn't see any need to _change_ the documentation, it just felt totally wasted hidden away on github where it's hard to find :)

So how's about this as a plan:

* First PR:
  * Move everything to `src/lib.rs` as I have done (and according to [recent best practice for crate documentation](https://blog.guillaume-gomez.fr/articles/2020-03-12+Guide+on+how+to+write+documentation+for+a+Rust+crate)
  * But also preserve the following things in the `README.md`: the format list, the open+save example, the fractal example (both code and image)
* Second and subsequent PRs:
  * Drop the `Pixel` stuff
  * Build up documentation on individual types for `GenericImage`/`ImageBuffer`/`DynamicImage` (probably by moving stuff from the top-level `src/lib.rs` and replacing with a concise paragraph explaining their relationships)
  * I eventually intend to provide some documentation for those various items which currently are missing documentation and therefore provide errors during `cargo doc` though this seems less urgent than getting the top-level docs right.
 
How does that sound?

Reply:
Oh, one more thing.

Another advantage of the docs being on `docs.rs` rather than on github is, of course, the ability to do the intra-crate links. I know this is obvious, but I wanted to explicitly state it's really useful to see a brief code example using (for instance) `GenericImage` and then be able to click a link to drill down into the documentation and find out about all the detailed methods. I suspect this is the workflow most newbies would initially follow when working out how to use this crate.

Reply:
> * Specifically, the code examples are the right level of complexity and I think are really helpful for newbies to the crate.

In the Readme, I do agree. That would also be the number one thing to introduce into the high-level documentation, it could even be worth duplicating for that purpose (it's really annoying that one can't use `include_str!` in the `#![doc]` macro as that would save us this duplication). There are a few other parts of the library that don't have quite enough code examples.

> * I think the explanation of the roles of GenericImage/ImageBuffer/DynamicImage is good and belongs in the top level documentation too. Yes, it could be shorter, but I don't see this as urgent.

I didn't meant to suggest removing the sections entirely, just that the code definitions is basically a duplicate of what rustdoc generates for the actual items and linking would be more apt and informative as it also allows browsing all of the related items. A short overview of their role is definitely relevant to the main documentation.

> * I think the table of formats is actually beneficial and probably belongs in the top-level documentation. I think it shows the number one feature of this library which is to support multiple image formats. (I'm not sure if you'd agree that's the number one feature but I assume it's why a lot of people want to use image).

That's probably correct, it is a good and popular reason to use `image`. Maybe it is worth to have the feature overview matrix and not just the list of crates. Although it really should be the module documentation that expands on this in-depth, for example an entry `all` obviously depends on the format. (Not that this is directly relevant for the PR but the first column were to actually link to the modules..?)

> * I think it would be shame to split information across the README.md and src/lib.rs. I think it's useful having a single page with all the information.

Not necessarily, the two files address different problems. Well structured documentation is far more valuable then a dump of _everything_ and part of structuring documentation is splitting it into well-ordered trees.


> So how's about this as a plan: [..]

I do like the overall goal of the plan. It might overblow my concerns, but I don't like having half-finished projects on the `master` branch, it is a hazard for making releases. It's no problem having a PR open and working on it for a longer time. You could also have a separate branch for improvements to docs until that repolish is finish, which I guess would improve the feedback cycles. The linked blog-post reads more like an introduction to good documentation practices with Rust (that said, even basics can have vastly superior results compared to other languages) with plenty of room for interpretation, and not in any way as an authoritative recommendation by wg-docs.

It should be said that the Readme _does_ exist [on `docs.rs`](https://docs.rs/crate/image/0.23.2), it's just not the redirection chosen when typing in the crate name alone and that leads to the exact confusion of first impression. A small mistake if you ask me but I don't think that `docs.rs` planned for being the first contact when one searches for crates, but a reference, although in practice it is quite often.

Reply:
OK. Thanks for taking the time to engage on this PR.

It sounds like you don't see a path by which incremental changes and improvements can be made to the docs, and that's fine, but then I don't think it's the project for me. I can't make the commitment to get a documentation improvement branch to a point where you'd necessarily be happy to merge to master, so I think someone else will have to take on that challenge in future.

As such I'm going to close this PR - but please could you consider merging #1181 instead? This is a very small change which will greatly improve ease of use for this module by newbies!